convert/equaler.go

```diff
@@ -1,23 +1,66 @@
 package convert
 
+import (
+	"runtime"
+	"strings"
+)
+
 // The Equaler interface allows object that implement it, to be compared
 type Equaler interface {
-	// Returns true if the Equaler object is the same as this object; otherwise false is returned.
-	// You need to convert the Equaler object using type assertions: https://golang.org/ref/spec#Type_assertions
+	// Equal returns true if the Equaler object is the same as this object;
+	// otherwise false is returned. You need to convert the Equaler object using
+	// type assertions: https://golang.org/ref/spec#Type_assertions
 	Equal(Equaler) bool
+	// EqualValue does the same as Equal but should ignore meta values like the
+	// version of an object or times when it was created, last updated or
+	// deleted.
+	EqualValue(Equaler) bool
+}
+
+// EqualValue is an internal function and shouldn't be called from the outside
+// but it needs to be exported for some calling logic that is embedeed into the
+// CallEqualByName and IsCallerEqualValue functions.
+func EqualValue(obj Equaler, other Equaler) bool {
+	return obj.EqualValue(other)
+}
+
+// CascadeEqual is a helper function that you can use to automatically pick the
+// correct equal function (Equal/EqualValue) depending on how the current equal
+// function is called.
+//
+// CascadeEqual calls obj.EqualValue(other) if the call to the calling function
+// ends was made from a function that ends with ".EqualValue"; otherwise
+// obj.Equal(other) is returned.
+func CascadeEqual(obj Equaler, other Equaler) bool {
+	pc, _, _, ok := runtime.Caller(2)
+	if !ok {
+		return false
+	}
+	if strings.HasSuffix(runtime.FuncForPC(pc).Name(), ".EqualValue") {
+		// We use this indirection so that CascadeEqual can operator
+		// successfully when called again.
+		return EqualValue(obj, other)
```

Comment:
If `obj.EqualValue` calls `CascadeEqual` internally, then it would see that the parent function was `CascadeEqual` and that doesn't help us in deciding on what to call next `Equal` or `EqualValue`. But if we have another layer in between the function name inspection above will evaluate to true on cascading calls to `CascadeEqual` and always choose the right equal method.

Consider an object that is made out of these sub/subsub objects:

`obj: {sub: {subsub:"foo"}}`

And now let's look at a call stack for `Equal` and `EqualValue` top to bottom:

**Equal():**
```go
obj.Equal(other)
convert.CascadeEqual(obj.sub, other.sub)
obj.sub.Equal(other.sub)
convert.CascadeEqual(obj.sub.subsub, other.sub.subsub)
obj.sub.subsub.Equal(other.sub.subsub)
```

**EqualValue():**
```go
obj.EqualValue(other)                       // overwrite values on obj from other
obj.Equal(other)                            // re-use already implemented Equal()
convert.CascadeEqual(obj.sub, other.sub)    // looks up 2 frames back (skip = 2) and
                                            // finds that it was EqualValue() that
                                            // triggered a call to Equal()
convert.EqualValue(obj.sub, other.sub)      // artificial layer
obj.sub.EqualValue(other.sub)               // overwrite to prepare and call Equal()
obj.sub.Equal(other.sub)
convert.CascadeEqual(obj.sub.subsub, other.sub.subsub) // look two frames back and
                                                       // voila: EqualValue() is
                                                       // found
convert.EqualValue(obj.sub.subsub, other.sub.subsub)   // artificial layer
obj.sub.subsub.EqualValue(other.sub.subsub)
obj.sub.subsub.Equal(other.sub.subsub)
```